paderbox/utils/iterable.py

```diff
@@ -0,0 +1,92 @@
+import collections
+import itertools
+import operator
+
+from typing import Callable, Any, Hashable, Optional, Union, Iterable
+
+
+__all__ = [
+    'groupby',
+]
+
+
+def zip(*iterables, strict=False):
+    """
+    Python implementation of strict zip, introduced in Python 3.10.
+    https://peps.python.org/pep-0618/
+    """
+    if not iterables:
+        return
+    iterators = tuple(iter(iterable) for iterable in iterables)
+    try:
+        while True:
+            items = []
+            for iterator in iterators:
+                items.append(next(iterator))
+            yield tuple(items)
+    except StopIteration:
+        if not strict:
+            return
+    if items:
+        i = len(items)
+        plural = " " if i == 1 else "s 1-"
+        msg = f"zip() argument {i+1} is shorter than argument{plural}{i}"
+        raise ValueError(msg)
+    sentinel = object()
+    for i, iterator in enumerate(iterators[1:], 1):
+        if next(iterator, sentinel) is not sentinel:
+            plural = " " if i == 1 else "s 1-"
+            msg = f"zip() argument {i+1} is longer than argument{plural}{i}"
+            raise ValueError(msg)
+
+
+def identity(x):
+    return x
+
+
+def groupby(
+        iterable,
+        key: Optional[Union[Callable[[Any], Hashable], str, int, Iterable]] = None,
+):
+    """
+    A non-lazy variant of `itertools.groupby` with advanced features.
+
+    Args:
+        iterable: Iterable to group
+        key: Determines by what to group. Can be:
+            - `None`: Use the iterables elements as keys directly
+            - `callable`: Gets called with every element and returns the group
+                key
+            - `str`, or `int`: Use `__getitem__` on elements in `iterable`
+                to obtain the key
+            - `Iterable`: Provides the keys. Has to have the same length as
+                `iterable`.
+
+    Examples:
+        >>> groupby('ab'*3)
+        {'a': ['a', 'a', 'a'], 'b': ['b', 'b', 'b']}
+        >>> groupby(range(10), lambda x: x%2)
+        {0: [0, 2, 4, 6, 8], 1: [1, 3, 5, 7, 9]}
+        >>> groupby(({'a': x%2, 'b': x} for x in range(3)), 'a')
+        {0: [{'a': 0, 'b': 0}, {'a': 0, 'b': 2}], 1: [{'a': 1, 'b': 1}]}
+        >>> groupby('abc', ['a'])
+        Traceback (most recent call last):
+            ...
+        ValueError: zip() argument 2 is longer than argument 1
+    """
+    value_getter = identity
+    if callable(key):
+        key_fn = key
+    elif key is None:
+        key_fn = identity
+    elif isinstance(key, (str, int)):
+        key_fn = operator.itemgetter(key)
+    else:
+        key_fn = operator.itemgetter(1)
+        value_getter = operator.itemgetter(0)
+        iterable = zip(key, iterable, strict=True)
```

Comment:
`iterable = zip(iterable, key, strict=True)` Produces a more natural error message.
Currently, `key_fn` and `value_getter` are swapped